Step02/gimun

wjdrlans5000 opened the pull request:
1차 pr

wjdrlans5000 committed:
feat : 로또 각 기능별 클래스 생성

wjdrlans5000 committed:
test : generate 메서드  테스트코드 작성

wjdrlans5000 committed:
feat : getter 추가

wjdrlans5000 committed:
feat : 입력한 금액으로 로또 장수 계산 및 잔액반환

입력한 금액을 기준으로 장당 천원씩 장수 계산 및 남은금액 반환, 구매금액이 천원 미만일 경우 IllegalArgumentException throw

wjdrlans5000 committed:
refactor : 로또생성 테스트 notnull인지 테스트케이스 추가

wjdrlans5000 committed:
refactor : 로또 당첨번호 지정

wjdrlans5000 committed:
refactor : 로또생성 수정

해당 기능에서 보너스 번호 생성이 불필요하여 제거

wjdrlans5000 committed:
refactor : 당첨번호 지정 수정

보너스번호까지 하나의 리스트로 받도록 수정

wjdrlans5000 committed:
feat : 당첨확인 기능 개발

wjdrlans5000 committed:
refactor : 테스트 메서드 명 한글로 변경

wjdrlans5000 committed:
feat : 당첨통계 기능 개발 및 테스트

wjdrlans5000 committed:
refactor : 로또 수동생성을 위해 세터추가

wjdrlans5000 committed:
refactor : 주석 및 import 정리

wjdrlans5000 committed:
feat : 입출력 메인메서드 작성

구매 금액 입력
로또 당첨 번호 입력
로또 구매 내역 출력
구매한 로또 번호 출력
당첨 내역 출력
수익률 출력

ces518 commented on STEP_02/src/me/gimun/BuyLotto.java:
```diff
@@ -0,0 +1,25 @@
+package me.gimun;
+
+public class BuyLotto {
+    private int lottoCnt;
+    private int balance;
+
+    public BuyLotto(int amount) {
+        if(amount < 1000){
+            throw new IllegalArgumentException();
+        }
+
+        this.lottoCnt = (int) Math.floor(amount/1000);
+        this.balance = amount%1000;
+    }
+
+    public int getLottoCnt() {
+        return lottoCnt;
+    }
+
+    public int getBalance() {
+        return balance;
+    }
+
+
+}
```
BuyLotto 라는 클래스가 하는 역할이 어떤건가요 ?
단순히 돈을 받아서 거스름돈을 반환하는 역할인것 같은데 BuyLotto 라는 클래스명과 어울리지 않는것 같습니다.

프로그램 EntryPoint 에 작성하신 코드를 보면 차라리 이런 클래스보다는 Store 라는 개념을 정의하고, Store 가 돈을 입력 받으면 로또 복권과 거스름돈을 반환해주는 방식이 더 낫지 않을까요

ces518 commented on STEP_02/src/me/gimun/BuyLotto.java:
```diff
@@ -0,0 +1,25 @@
+    private int balance;
+
+    public BuyLotto(int amount) {
+        if(amount < 1000){
```
1000 이라는 리터럴 상수가 의미 하는 값이 뭘까요 ?
현재 여러 곳에서 사용하고 있고 의미를 파악하기 힘든것 같은데 좋은 방법은 없을까요 ?

ces518 commented on STEP_02/src/me/gimun/ConfirmWinning.java:
```diff
@@ -0,0 +1,46 @@
+                }
+            }
+        }
+        switch (winNumbers){
+            case 6:
+                result = "1등";
+                break;
+            case 5:
+                if(isBonus){
+                    result = "2등";
+                }else{
+                    result = "낙첨";
+                }
+                break;
+            case 4:
+                result = "3등";
+                break;
+            case 3:
+                result = "4등";
+                break;
+            default:
+                result = "낙첨";
+                break;
+        }
+
+        return result;
```
1등, 2등 이런 부분은 View 에 해당하는 로직이라 당첨 결과를 확인하는 BizLogic 과 분리되어야 할것 같습니다

ces518 commented on STEP_02/src/me/gimun/LottoGenerate.java:
```diff
@@ -0,0 +1,33 @@
+        Set<Integer> set = new HashSet<>();
+        Random random = new Random();
+        while (set.size() != 6){
+            set.add(random.nextInt(45 ) + 1);
```
로또번호 생성시 유효 범위에 대한 책임이 로또 번호 생성기가 가져야 하는걸까요 ?
로또번호 라는 도메인 객체를 생성하고 해당 객체 내부에서 체크하는것이 좋을것 같아요

ces518 commented on STEP_02/src/me/gimun/LottoGenerate.java:
```diff
@@ -0,0 +1,33 @@
+    public void generate(){
+        Set<Integer> set = new HashSet<>();
+        Random random = new Random();
+        while (set.size() != 6){
```
6이라는 리터럴 상수가 가지는 의미를 알 수 있으면 좋을 것 같습니다

ces518 commented on STEP_02/src/me/gimun/LottoGenerate.java:
```diff
@@ -0,0 +1,33 @@
+
+    public void generate(){
+        Set<Integer> set = new HashSet<>();
+        Random random = new Random();
```
로또 번호의 범위는 정해져 있는데, 생성시 굳이 매번 랜덤 객체를 이용해서 생성할 필요가 있을까요 ?
로또 범위의 번호를 어딘가에 미리 생성해두고 (캐싱) 해당 번호들 중에 랜덤하게 뽑아오는게 더 낫지않을까요 ?

ces518 commented on STEP_02/src/me/gimun/SettingWinningNumber.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class SettingWinningNumber {
```
SettingWinningNumber라는 네이밍이 잘 와닿지 않는데요
역할에 맞는 더 좋은 네이밍을 사용하는 좋을것 같습니다

ces518 commented on STEP_02/src/me/gimun/SettingWinningNumber.java:
```diff
@@ -0,0 +1,33 @@
+    private List<Integer> winningLotto;
+    private int winningBonusNum;
+
+    public SettingWinningNumber() {
+        this.winningLotto = new ArrayList<Integer>();
+    }
+
+    public void winningNumSet(List<Integer> lotto){
+        //로또 길이가 7이 아니면 에러
+        if(lotto.size() != 7){
+            throw new IllegalArgumentException();
+        }
+
+        this.winningBonusNum = lotto.get(lotto.size()-1);
+        lotto.remove(lotto.size()-1);
+        this.winningLotto = lotto;
+
+    }
```
생성자의 로직과 winningNumSet 은 항상 동시에 일어나야 하는것 같은데, 굳이 분리한 이유가 있을까요 ?
생성자로 받아 lotto 숫자 목록이 필수 값임을 명시하고, 로직을 수행하는 방식이 더 나을것 같습니다

ces518 commented:
전체적인 도메인 설계에 대해서 좀 더 고민을 해보시는게 좋을것 같고, 객체간의 협력을 통한 동작보다는 단순히 명령형으로 동작하도록 작성된 코드가 많습니다.

또한 하드코딩된 부분, 객체간의 책임 분리 등 어떻게 하면 더 이해하기 쉽고, 변경에 유연하게 대처할 수 있을지 고민해보시면 좋을것 같습니다

wjdrlans5000 committed:
refactor : SettingWinningNumber 네이밍 변경 및 생성자에서 winningNumSet 로직 수행

SettingWinningNumber -> SetWinNum(당첨번호설정)
winningNumSet 로직 생성자와 통합
테스트 코드 및 메인 출력코드 SetWinNum 수정

wjdrlans5000 committed:
refactor : BuyLotto 클래스명 변경 및 상수정의

BuyLotto 클래스명 LottoStore로 변경하여 돈을 입력받으면 구매한 로또 장수와 거스름돈을 계산
로또 가격 상수정의

wjdrlans5000 committed:
refactor : View에 해당하는 로직 분리

당첨결과 View에 해당하는 로직 메서드 분리

wjdrlans5000 committed:
feat : 로또번호 도메인 객체 생성

해당 도메인 객체에서 각 로또번호 생성 및 Validation 체크, 로또범위 미리 생성등의 로직 수행

wjdrlans5000 committed:
test : 로또번호 도메인 객체 테스트

wjdrlans5000 committed:
refactor : 로또 사이즈 상수설정 및 generate 메서드 수정

6 상수 로또사이즈를 나타내는 상수로 설정
generate 메서드에서 LottoNumber 도메인에서 캐싱해둔 로또 범위와 동일한 인스턴스를 사용하여 로또 자동생성 하도록 수정

wjdrlans5000 committed:
refactor : 각 클레스 수정 및 추가에 의한 출력 메소드 변경

wjdrlans5000 commented:
2차 pr

ces518 commented on STEP_02/src/me/gimun/LottoGenerate.java:
```diff
@@ -0,0 +1,25 @@
+
+        List<LottoNumber> generatedNumber = lottoNumbers.subList(0, LOTTO_SIZE);
+
+        Collections.sort(generatedNumber, new Comparator<LottoNumber>() {
+            @Override
+            public int compare(LottoNumber num1, LottoNumber num2) {
+                return Integer.valueOf(num1.getNumber()).compareTo(Integer.valueOf(num2.getNumber()));
+            }
+        });
```
Comparator 를 익명클래스로 구현한 이유가 있을까요 ?

ces518 commented on STEP_02/src/me/gimun/LottoStore.java:
```diff
@@ -0,0 +1,26 @@
+            throw new IllegalArgumentException("입력하신 금액이 로또 금액보다 작습니다.");
+        }
+
+        this.lottoCnt = (int) Math.floor(amount/this.LOTTO_AMOUNT);
+        this.balance = amount%this.LOTTO_AMOUNT;
```
Math.floor 를 사용하지 않아도 될 것 같아요

ces518 commented on STEP_02/src/me/gimun/LottoStore.java:
```diff
@@ -0,0 +1,26 @@
+package me.gimun;
+
+public class LottoStore {
+    private final int LOTTO_AMOUNT = 1000;
```
로또 장당 가격이 변경되면 LottoStore 를 수정해야 할 것 같은데 다른 방법은 없을까요 ?

ces518 commented on STEP_02/src/me/gimun/SetWinNum.java:
```diff
@@ -0,0 +1,30 @@
+            throw new IllegalArgumentException();
+        }
+
+        this.winningBonusNum = lotto.get(lotto.size()-1);
```
현재 구현을 보면 SetWinNum 이 보너스 번호를 산정하는 비즈니스를 책임 지고 있는데 역할의 분리가 잘 이뤄진것 같지 않습니다
비즈니스가 변경되서 보너스 번호가 로또의 가장 앞 번호로 변경된다면 어떻게 해야할까요 ?

ces518 commented on STEP_02/src/me/gimun/SetWinNum.java:
```diff
@@ -0,0 +1,30 @@
+        }
+
+        this.winningBonusNum = lotto.get(lotto.size()-1);
+        lotto.remove(lotto.size()-1);
+        this.winningLotto = lotto;
```
보너스 번호를 제외한 로또 번호를 추출하고 싶으셨던것 같은데 원본을 직접 수정하는 방식은 좋은 방식이 아닌것 같습니다
불변성을 지키는쪽이 좋을것 같아요

ces518 commented on STEP_02/src/me/gimun/domain/LottoNumber.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.List;
+import java.util.Objects;
+
+public class LottoNumber {
```
public / private / static 메소드가 현재 뒤죽박죽 섞여있는데요
컨벤션에 맞춰 정리가 필요해보입니다

ces518 commented on STEP_02/src/me/gimun/domain/LottoNumber.java:
```diff
@@ -0,0 +1,47 @@
+        }
+    }
+
+    public static List<LottoNumber> getCACHE() {
+        return CACHE;
+    }
```
Cache 라는 컬렉션 객체의 참조를 그대로 반환하셨는데, 이러면 외부에서 조작이 가능해지고 이로 인해 의도치 않은 동작이 발생할 수 있을것 같습니다 좋은 방법 없을까요?

ces518 commented on STEP_02/src/me/gimun/domain/LottoNumber.java:
```diff
@@ -0,0 +1,47 @@
+    private static final List<LottoNumber> CACHE = new ArrayList<>();
+
+    static {
+        for (int i = LOTTO_NUMBER_LOWER_BOUND; i <= LOTTO_NUMBER_UPPER_BOUND; i++) {
+            CACHE.add(new LottoNumber(i));
+        }
```
Java Stream API 를 활용해보는건 어떨까요 ?

ces518 commented on STEP_02/test/me/gimun/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,28 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class LottoNumberTest {
```
잘못된 로또번호 생성에 대한 테스트는 없는것 같아요

ces518 commented on STEP_02/test/me/gimun/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,28 @@
+
+class LottoNumberTest {
+
+    @Test
+    void 로또번호객체생성() {
+        List<LottoNumber> lotto = new ArrayList<>();
+        int jj = 0;
+        while (jj != 6){
+            lotto.add(LottoNumber.valueOf(jj));
+            jj++;
+        }
+
+        assertNotNull(lotto);
+
+        for (LottoNumber lottoNum : lotto) {
+            System.out.println(lottoNum.getNumber());
+        }
+    }
```
테스트가 잘못된것 같은데 다시 한번 고민해보시면 좋을것 같아요

ces518 commented on STEP_02/src/me/gimun/LottoGenerate.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.*;
+
+public class LottoGenerate {
+    private static final int LOTTO_SIZE = 6;
+
+    public static List<LottoNumber> generate() {
```
LottoGenerate 를 class 로 사용 하셨는데 만약 로또 생성방식이 변경되면 어떻게 될까요 ?

wjdrlans5000 committed:
refactor : 로또스토어 수정 및 로또 enum 추가

math.floor 제거
로또 장당 가격을 enum으로 설정하여 가격이 변경되더라도 LottoStore 클래스를 직접 수정하지 않도록 수정

wjdrlans5000 committed:
refactor : 로또 제너레이터 정렬 기능 수정

기존 Comparator를 익숙한 사용방식인 익명클래스로 구현하였었는데 람다표현식으로 간단히 표현

wjdrlans5000 committed:
refactor : 로또번호 Cache 컬렉션 객체 참조 반환 수정

Cache 컬렉션 객체의 참조를 직접 반환하지않고 새 ArrayList 객체에 복사하여 반환

wjdrlans5000 committed:
refactor : 로또번호 Cache 생성방식 수정

기존 for문 을 사용한 방식에서 Stream API를 사용하여 1~45까지의 숫자 캐싱

wjdrlans5000 committed:
refactor :  당천번호 생성과 보너스번호 생성 로직 분리

보너스번호 생성 로직 분리 및 관련 테스트, 입출력 코드 변경

wjdrlans5000 committed:
refactor :  로또번호 클래스 메서드 정리

wjdrlans5000 committed:
refactor :  로또번호생성 테스트코드 수정

잘못된 테스트 수정
잘못된 로또번호 생성에 대한 테스트 추가

wjdrlans5000 committed:
refactor : 캐시생성 람다표현식 수정

wjdrlans5000 committed:
refactor : 로또번호 자동생성 interface 변경 및 상속

generate() > interface의 default 메서드로 생성하여 로또생성방식의 변경이 필요할 경우 해당 메서드 오버라이딩하여 사용할수 있도록 수정
Comparator 수정 > comparingInt 및 메서드레퍼런스 사용
로또번호 자동생성 관련 테스트코드 수정

wjdrlans5000 commented:
3차pr

ces518 commented on STEP_02/src/me/gimun/LottoStore.java:
```diff
@@ -0,0 +1,26 @@
+package me.gimun;
+
+public class LottoStore {
+    private final int LOTTO_AMOUNT = Lotto.LOTTO_AMOUNT.getAmount();
```
기존에 적용되어 있던 상수를 ENUM 으로 변경 하셨는데요
이 또한 상수로 처리되는 부분이라 로또 가격의 변경이 필요하면 LottoStore 에 수정이 필요할것 같습니다
예를 들어 로또를 특정 장수이상 구매하게되면 할인가가 들어간다거나 하게 되면 어떤식으로 대응을 할 수 있을까요 ?

ces518 commented on STEP_02/src/me/gimun/LottoNumGenerate.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Comparator;
+import java.util.List;
+
+public interface LottoNumGenerate {
+    int LOTTO_SIZE = 6;
+
+    default List<LottoNumber> generate() {
+        List<LottoNumber> lottoNumbers = new ArrayList<>(LottoNumber.getCACHE());
+        Collections.shuffle(lottoNumbers);
+
+        List<LottoNumber> generatedNumber = lottoNumbers.subList(0, LOTTO_SIZE);
+        generatedNumber.sort(Comparator.comparingInt(LottoNumber::getNumber));
+        return generatedNumber;
+    }
+}
```
인터페이스를 통해 느슨한 관계를 형성해주신것 좋습니다.

ces518 replied:
default 메소드를 활용하신건 좋으나 default 메소드의 본래 목적과 다르게 사용을 하셨는데요
DefaultLottoNumGenerator 라는 클래스를 두고 해당 클래스가 기본 구현을 하게 하는 방식이 더 좋을것 같습니다

ces518 commented on STEP_02/test/me/gimun/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,46 @@
+        System.out.println(setLotto);
+    }
+
+    @Test
+    void 잘못된로또번호생성() {
+        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, ()-> new LottoNumber(0));
+        String message = exception.getMessage();
+
+        assertEquals("LottoNumber가 유효하지 않습니다.",message);
+
+    }
```
잘못된 로또 번호 생성에 대한 테스트를 작성해주셨는데 제가 느끼기엔 반쪽자리 테스트 같습니다
로또 번호의 유효범위가 1~46 이라면 이 범위를 벗어나는 경우 잘못된 로또번호가 아닌가요 ?
현재 테스트만 보았을때 0일 경우에만 잘못된 번호라고 인지하게 될 것 같아요

ces518 replied:
테스트 작성시 성공하는 테스트도 중요하지만 실패케이스를 얼마나 잘 캐치하느냐가 중요합니다

ces518 commented on STEP_02/src/me/gimun/domain/LottoNumber.java:
```diff
@@ -0,0 +1,49 @@
+public class LottoNumber {
+    private static final int LOTTO_NUMBER_LOWER_BOUND = 1;
+    private static final int LOTTO_NUMBER_UPPER_BOUND = 45;
+    private static final List<LottoNumber> CACHE = new ArrayList<>();
+
+    static {
+        IntStream stream = IntStream.rangeClosed(LOTTO_NUMBER_LOWER_BOUND, LOTTO_NUMBER_UPPER_BOUND);
+        stream.forEach(i -> CACHE.add(new LottoNumber(i)));
+    }
```
구현부를 보니 반쪽자리 스트림 코드인데 이런 식이면 for 문을 사용할때와 큰 차이가 없는것 같습니다

ces518 commented on STEP_02/src/me/gimun/ConfirmWinning.java:
```diff
@@ -0,0 +1,54 @@
+        String result = "";
+        int winNumbers = 0;
+        boolean isBonus = false;
+        for (int i = 0; i < buyLotto.size(); i++) {
+            for (int j = 0; j < winningLotto.size(); j++) {
+                if (buyLotto.get(i).getNumber() == winningLotto.get(j)) {
+                    winNumbers++;
+                }
+                else if(buyLotto.get(i).getNumber() != winningLotto.get(j)
+                            && buyLotto.get(i).getNumber() == bonus){
+                    isBonus = true;
+                }
+            }
+        }
```
로직을 개선할 수 있을것 같은데 고민해보시면 좋을것 같아요

ces518 commented:
아직 각 기능들이 제대로된 모듈화가 되지 않은 것 같습니다
EntryPoint 인 main 메소드를 살펴보면 명령형 코드에 더 가까운데 고민을 좀 더 해보시는게 어떨까요 ?

wjdrlans5000 committed:
refactor : LottoNumGenerate 수정

default 메소드 제거하고 DefaultLottoNumGenerator 클래서에서 해당 인터페이스 기본 구현
관련 테스트 수정

wjdrlans5000 committed:
refactor : 로또번호 캐싱 스트림코드 수정

foreach를 사용하지않고 mapToObj 사용하여 내부적으로 값을 처리하도록 수정
잘못된 로또번호 생성에 대한 테스트 보완

wjdrlans5000 committed:
refactor : 로또스토어 할인추가

로또스토어 생성자 오버로딩(할인로직 람다 파라미터 추가)
관련 테스트코드 작성

wjdrlans5000 committed:
refactor : 로또당첨 로직 수정

HashMap의 getOrDefault 사용하여 for문의 반복횟수를 줄여봄..

wjdrlans5000 committed:
refactor : 메인메서드 각 모듈별 메서드 분리

wjdrlans5000 commented:
4차 pr 입니다.  
로또당첨 로직에 대해서는 for문 반복횟수를 줄이고자 해봤는데 방향이 맞는지 모르겠네요.. 사이즈가 큰 list가 아니라서 그런지 성능 차이는 개선전과 별 차이가 없는거 같아서요 피드백 부탁드립니다.

ces518 commented on STEP_02/src/me/gimun/Main.java:
```diff
@@ -0,0 +1,106 @@
+import java.util.*;
+
+public class Main {
+    public static void main(String[] args) throws IOException {
+        //구매급액 입력
+        System.out.println("로또를 구매하실 금액을 입력하세요.(장당 천원)");
+        LottoStore lottoStore = new LottoStore(inputStream());
+
+        //로또 당첨번호
+        SetWinNum setWinNum = setWinNumBonus();
+
+        //로또 구매 내역 출력
+        System.out.println("로또 구매 장수 : " + lottoStore.getLottoCnt());
+        System.out.println("잔액 : " + lottoStore.getBalance());
+
+        //구매한 로또 번호
+        List<List<LottoNumber>> buyLottoList = buyLottoList(lottoStore);
+
+        //당첨 내역 출력
+        System.out.println("당첨번호 : " + setWinNum.getWinningLotto());
+        System.out.println("보너스번호 : " + setWinNum.getWinningBonusNum());
+
+        //당첨통계설정
+        List<String> list = confirmWinning(buyLottoList, setWinNum);
+
+        //당첨통계출력
+        StatisticWinning statisticWinning = new StatisticWinning();
+        statisticWinning.calculate(list, lottoStore.getLottoCnt());
+        System.out.println("수익률(%) : " +statisticWinning.getRateOfReturn());
+
+    }
```
흠.. 의미 전달이 잘 안된것 같은데요 
단순히 메소드로 추출하라는게 아니라 모듈화의 의미를 잘 생각해보시는게 좋을것 같습니다
현재도 메소드로 찢어만 놨을뿐 명령형 코드로 구현되어 있습니다

ces518 commented on STEP_02/src/me/gimun/Main.java:
```diff
@@ -0,0 +1,106 @@
+            set.add(inputStream());
+            ii++;
+        }
+        List<Integer> winningLotto = new ArrayList<Integer>();
+        winningLotto.addAll(set);
+        Collections.sort(winningLotto);
+
+        SetWinNum setWinNum = new SetWinNum(winningLotto);
+        System.out.println("보너스번호를 입력 하세요.");
+        //보너스번호
+        setWinNum.setBounsNum(inputStream());
+        return setWinNum;
+    }
```
중복되지 않은 숫자 6개로 구성되어 있고, 정렬된 상태로 존재해야 한다.
이 로직이 Main 클래스에 있는게 맞을까요 ?

ces518 commented on STEP_02/src/me/gimun/Main.java:
```diff
@@ -0,0 +1,106 @@
+        List<String> list = new ArrayList<String>();
+        int i = 0;
+        for (List<LottoNumber> numbers: buyLottoList) {
+            String result = confirmWinning.confirmLotto(setWinNum.getWinningLotto(), setWinNum.getWinningBonusNum(),numbers);
```
SetWinNum 이 당첨번호와 보너스번호 모두 가지고 있는데 따로따로 꺼내서 호출하신 이유가 있나요

ces518 commented on STEP_02/src/me/gimun/SetWinNum.java:
```diff
@@ -0,0 +1,32 @@
+    private List<Integer> winningLotto;
+    private int winningBonusNum;
+
+    public SetWinNum(List<Integer> lotto) {
+        this.winningLotto = new ArrayList<>();
+        //로또 길이가 6이 아니면 에러
+        if(lotto.size() != 6){
+            throw new IllegalArgumentException("Lotto 길이가 6이 아닙니다.");
+        }
+
+        this.winningLotto = lotto;
+    }
```
이전에 정의해둔 LottoNumber 라는 도메인 클래스와 중복되지 않나요 ?

ces518 commented on STEP_02/src/me/gimun/SetWinNum.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class SetWinNum {
```
SetWinNum 이라는 네이밍 보다는 당첨번호 라는 의미를 명확하게 전달할 수 있는 네이밍이 좋을것 같습니다

ces518 commented on STEP_02/src/me/gimun/LottoStoreIf.java:
```diff
@@ -0,0 +1,7 @@
+package me.gimun;
+
+public interface LottoStoreIf {
+
+    int LottoDisCount(int amount);
+
+}
```
해당 클래스가 수행하는 역할은 무엇인가요

ces518 commented on STEP_02/src/me/gimun/LottoStore.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.Optional;
+
+public class LottoStore {
+    private int lotto_amount = 1000;
```
이전에도 드렸던 코멘트 같은데요
로또 장당 구매가격이 변경되면 LottoStore 클래스가 수정이 필요합니다.
이런 경우 어떤 방식을 취하는것이 좋을까요 ?

ces518 commented on STEP_02/src/me/gimun/LottoStore.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public LottoStore(int amount, LottoStoreIf p) {
+        if(amount < this.lotto_amount){
+            throw new IllegalArgumentException("입력하신 금액이 로또 금액보다 작습니다.");
+        }
```
위 생성자와 중복코드인것 같습니다

ces518 commented on STEP_02/src/me/gimun/LottoStore.java:
```diff
@@ -0,0 +1,40 @@
+                (val)-> this.lotto_amount = val.LottoDisCount(amount)
+        );
+
+        this.lottoCnt = amount/this.lotto_amount;
+        this.balance = amount%this.lotto_amount;
```
이부분도 동일합니다

ces518 commented on STEP_02/test/me/gimun/DefaultLottoNumGeneratorTest.java:
```diff
@@ -0,0 +1,27 @@
+
+class DefaultLottoNumGeneratorTest {
+
+    @Test
+    void 로또생성() {
+
+        List<LottoNumber> lotto = new DefaultLottoNumGenerator().generate();
+
+        assertNotNull(lotto.get(0).getNumber());
+
+        List<Integer> setLotto = new ArrayList<>();
+        for (LottoNumber lottoNum : lotto) {
+            setLotto.add(lottoNum.getNumber());
+        }
+
+        System.out.println(setLotto);
+    }
```
테스트코드만 봐서는 로또넘버생성기의 기본 전략이 무엇인지 알 수 없는데요
기존 요구사항은 **1~45 범위의 수로 구성되어있고, 중복되지 않으며 정렬되어 있어야한다.** 인데 테스트코드만 봤을땐 해당 요구사항을 알 수 없습니다
테스트코드의 목적은 기능의 검증도 있지만 문서화의 역할도 수행하는데요, 위 테스트코드만 보았을땐 요구사항이 변경되거나 버그가 발생해도 코드 레벨에서 발견되기 어려운것 같습니다

ces518 commented on STEP_02/src/me/gimun/ConfirmWinning.java:
```diff
@@ -0,0 +1,69 @@
+//            }
+//        }
+
+        HashMap<Integer, Object> hashMap = new HashMap<>();
+        for (int i = 0; i< winningLotto.size(); i++) {
+            hashMap.put(winningLotto.get(i),i+1);
+        }
```
맵을 사용할 이유가 있을까요 ?

ces518 commented on STEP_02/src/me/gimun/ConfirmWinning.java:
```diff
@@ -0,0 +1,69 @@
+        }
+
+        for (LottoNumber lotto : buyLotto) {
+            int key = (int) hashMap.getOrDefault(lotto.getNumber(),0);
+            if(key != 0){
```
이런 방식은 좋지 않은 패턴입니다. default 값을 지정해 주었지만 해당 값이 null 과 동일한 역할을 하고 있는데요
이런 경우엔 오히려 null 을 사용하는 방법이 훨씬 낫습니다
**비즈니스 널** 에 관해 찾아보시면 좋을 것 같습니다

ces518 reviewed (changes requested):
PR 을 진행할수록 점진적으로 설계에 대해 고민하는 시간이 많아져야 하는데 비해 코드를 봤을때 고민하는 시간이 적었다는게 보여집니다.
단순 구현이 아닌 좋은 설계를 하기위한 고민을 많이 해보시는게 좋을 것 같습니다
